Added nextPow2 and truncPow2

JackStouffer opened the pull request:
https://issues.dlang.org/show_bug.cgi?id=6343

Added a new function to std.math `nextPow2`. The code here is adapted from the DDMD function, which is why it is different than the suggested code in the bug report. Also, this code is safe, as well as 11% faster.

ping @9il, you asked me to notify you when I made a math PR.

JackStouffer commented:
@CyberShadow In the doc generator, the prototype for this function is listed as

```
ulong nextPow2(
  const(ulong) val
) pure nothrow @nogc @safe;
```

which is clearly wrong.

CyberShadow commented:
DDox doesn't understand `size_t` because DMD resolves the alias and writes the resolved type in the JSON output. It is a known DDox limitation.

CyberShadow commented:
https://github.com/rejectedsoftware/ddox/issues/42

CyberShadow commented:
This function has no reason to use `size_t`, actually. `size_t` is tied to the address space size, and this function is not operating on addresses or sizes.

CyberShadow commented on std/math.d:
```diff
@@ -7291,3 +7292,85 @@ unittest
+
+unittest
+{
+    version(X86)
```
And this is just wrong. What about non-x86 32-bit architectures?

DmitryOlshansky replied:
Indeed e.g. `static if(size_t.sizeof == 4)`

CyberShadow commented on std/math.d:
```diff
@@ -7291,3 +7292,85 @@ unittest
+        2251799813685248, 4503599627370496, 9007199254740992,
+        18014398509481984, 36028797018963968, 72057594037927936,
+        144115188075855872, 288230376151711744, 576460752303423488,
+        1152921504606846976, 2305843009213693952, 4611686018427387904];
```
You could've at least used binary or hexadecimal base constants to make this list readable.

DmitryOlshansky replied:
Or use trivial CTFE to generate x^^2 table.

CyberShadow commented:
Is this really the fastest way to implement this function?

What about `bsr`?

What about binary or?

9il commented:
You may want to use some function from http://dlang.org/phobos/core_bitop.html to optimise this.

9il commented:
Generalisation issues:

1. Signed integers.
2. Floating point numbers.
3. `nextUpPow2` + `nextDownPow2` - next down analog.

9il commented:
@CyberShadow could you please add `math` label?

CyberShadow commented:
My branchless implementation (using binary or):
https://github.com/CyberShadow/ae/blob/master/utils/math.d#L64-L78

DmitryOlshansky commented:
> My branchless implementation (using binary or):
> https://github.com/CyberShadow/ae/blob/master/utils/math.d#L64-L78

Like it

JackStouffer commented:
> My branchless implementation (using binary or):
https://github.com/CyberShadow/ae/blob/master/utils/math.d#L64-L78

That code has a different behavior than the current one, where currently `nextPow2(2) == 2` and your code is `nextPow2(2) == 4`. The bug report asks for the first one, which is preferred?

CyberShadow commented:
Well, the function's name *is* `nextPow2`. The behavior you describe is what I have as `roundUpToPowerOfTwo`.

JackStouffer commented:
> nextUpPow2 + nextDownPow2 - next down analog.

One thing at a time :)

9il commented:
> One thing at a time :)

OK, but names. something like `ceilPow2` and `floorPow2`. First we need to understand behavior for signed integers.

JackStouffer commented:
> 1. Signed integers.

I honestly can't think of a time where having the next negative power of two would be useful. And even if if did need it, I could always just `-nextPow2(abs(x))`. So I think having a template constraint that rejects anything under zero makes the most sense.

> 2. Floating point numbers.

Well, there are two options, make it so the template constraint rejects them, or round them. I think the first option is what most people would expect.

CyberShadow commented:
> So I think having a template constraint that rejects anything under zero

How do you plan on doing that? Did you mean a contract, or rejecting signed types?

> Well, there are two options, make it so the template constraint rejects them, or round them.

I'm not totally sure but I think the IEEE representation should allow doing this very efficiently.

JackStouffer commented:
> How do you plan on doing that? Did you mean a contract, or rejecting signed types?

I ment contract. I edited as soon as I realized my mistake.

> I'm not totally sure but I think the IEEE representation should allow doing this very efficiently.

Which part are you referring to? You mention both options in the part you quote.

CyberShadow commented:
Well, both. The sign is just a bit in IEEE, there's no two's-complement to worry about. And the rounding as well, don't you essentially just need to set the mantissa to 1 and increment the exponent?

JackStouffer commented:
Changed the code to address a lot of the concerns.

CyberShadow commented on std/math.d:
```diff
@@ -7291,3 +7292,59 @@ unittest
+Returns:
+    the next highest power of two after $(D val)
+*/
+T nextPow2(T)(T val) if (isIntegral!T && !isFloatingPoint!T)
```
`isIntegral!T` implies `!isFloatingPoint!T`, doesn't it?

JackStouffer replied:
You're right, the documentation made it seem otherwise.

CyberShadow commented on std/math.d:
```diff
@@ -7291,3 +7292,59 @@ unittest
+
+/**
+Gives the next highest power of two after $(D val). $(T) can be any non-floating
+point numerical type. If the input would result in an overflow, then zero is
```
"any non-floating point numerical type"? That's ambiguous, might imply e.g. `BigInt` or `char`... Why not just "built-in integral type"?

CyberShadow commented on std/math.d:
```diff
@@ -7291,3 +7292,59 @@ unittest
+returned.
+
+Params:
+    val = a non-floating point number
```
Should mention that it cannot be zero

JackStouffer commented:
@CyberShadow fixed

9il commented:
> Well, both. The sign is just a bit in IEEE, there's no two's-complement to worry about. And the rounding as well, don't you essentially just need to set the mantissa to 1 and increment the exponent?

Agreed. @JackStouffer Please add implementation for FP. So we can be sure about new fanction names.

CyberShadow commented:
LGTM.

> @JackStouffer Please add implementation for FP.

IEEE internals are non-trivial stuff, this might be better left for a separate PR.

JackStouffer commented:
Could someone remove the `issue 6343` from the title, apparently this is not what the bug reporter wanted.

9il commented:
> IEEE internals are non-trivial stuff, this might be better left for a separate PR.

It would be better to make single review both for FP and integers because we can lost some important point right now.

CyberShadow commented:
@JackStouffer Are you OK with writing an FP implementation? If not, @9il, you could make a PR to this PR's branch (https://github.com/JackStouffer/phobos/tree/pow2)

JackStouffer commented:
@CyberShadow I understand how FPs work, but manipulating them is beyond me.

CyberShadow commented:
@9il, can you contribute a FP implementation?

9il commented:
OK, I will create separate PR, but don't merge this. So, we be able to review them at once.

JackStouffer commented:
@9il ping.

9il commented:
Excuse me for delay.
So, it can be done easily with `frexp` and `ldexp`.
```D
import std.traits;
import std.math;

T nextPow2(T)(const T x)
    if (isFloatingPoint!T)
{
    if(!x.isFinite)
        return x;
    if(!x)
    	return x;
    int exp;
    auto y = frexp(x, exp);
    y = ldexp(cast(T)0.5, exp+1);
    y = copysign(y, x);
    return y;
}

unittest {
	import std.meta: AliasSeq;
	foreach(T; AliasSeq!(float, double, real))
	{
		enum T subNormal = T.min_normal / 2;

		assert(nextPow2(subNormal) == T.min_normal);

		assert(nextPow2(T(0.0)) == 0.0);

		assert(nextPow2(T(2.0)) == 4.0);
		assert(nextPow2(T(2.1)) == 4.0);
		assert(nextPow2(T(3.1)) == 4.0);
		assert(nextPow2(T(4.0)) == 8.0);
		assert(nextPow2(T(0.25)) == 0.5);

		assert(nextPow2(T(-2.0)) == -4.0);
		assert(nextPow2(T(-2.1)) == -4.0);
		assert(nextPow2(T(-3.1)) == -4.0);
		assert(nextPow2(T(-4.0)) == -8.0);
		assert(nextPow2(T(-0.25)) == -0.5);

		assert(nextPow2(T.max) == T.infinity);
		assert(nextPow2(-T.max) == -T.infinity);

		assert(nextPow2(T.infinity) == T.infinity);
		assert(nextPow2(T.init).isNaN);
	}
}
```

9il commented on std/math.d:
```diff
@@ -7291,3 +7292,58 @@ unittest
+T nextPow2(T)(T val) if (isIntegral!T)
+in
+{
+    assert(val >= 0, "val must be greater than zero");
```
`greater` or `greater or equal`?

JackStouffer commented:
@9il The integer version doesn't handle negative numbers but the floating point version does. Should the integer version be changed?

JackStouffer commented:
Also, would it make more sense if `nextPow2(-4.0) == -2.0` so `nextDownPow2(-4.0) == -8.0`?

9il commented:
> The integer version doesn't handle negative numbers but the floating point version does. Should the integer version be changed?

I think so.

> Also, would it make more sense if nextPow2(-4.0) == -2.0 so nextDownPow2(-4.0) == -8.0?

Probably this function would be used for amplitude rounding. So nextPow2(-4.0) == -8.0 looks more comfortable.

9il commented:
`nextDownPow2` can be replaced with `truncateDownPow2 = nextPow2 / 2`

JackStouffer commented:
@CyberShadow @9il The floating point implementation 9il wrote, modified slightly to follow the behavior of the integer implementation. Also, integer implementation now handles negative numbers correctly.

JackStouffer commented:
Ping.

9il commented on std/math.d:
```diff
@@ -7291,3 +7292,127 @@ unittest
+ */
+T nextPow2(T)(const T val) if (isIntegral!T)
+{
+    Unqual!T x;
```
No need `Unqual` here

9il replied:
> No need Unqual here

I was wrong, it is more useful with `Unqual` for type generic code.

9il commented on std/math.d:
```diff
@@ -7291,3 +7292,127 @@ unittest
+
+    assert(nextPow2(-2) == -4);
+    assert(nextPow2(-10) == -16);
+    assert(nextPow2(size_t.max) == 0);
```
please, add examples for min/max x signed/unsigned.

9il commented on std/math.d:
```diff
@@ -7291,3 +7292,127 @@ unittest
+        assert(nextPow2(T.infinity) == T.infinity);
+        assert(nextPow2(T.init).isNaN);
+    }
+}
```
Nitpick: add please blank line

9il commented:
other than that  LGTM

JackStouffer commented:
@9il updated.

9il commented on std/math.d:
```diff
@@ -7291,3 +7292,137 @@ unittest
+    {
+        enum T subNormal = T.min_normal / 2;
+
+        assert(nextPow2(subNormal) == T.min_normal);
```
```D
static if(subNormal)
    assert(nextPow2(subNormal) == T.min_normal);
```
subNormal can be `0` for some CPUs.

9il commented on std/math.d:
```diff
@@ -7291,3 +7292,137 @@ unittest
+{
+    Unqual!T x;
+
+    if (val < 0)
```
```D
x = val;
static if(isSigned!T)
    x = abs(x);
```

JackStouffer commented:
@9il fixed

9il commented on std/math.d:
```diff
@@ -7291,3 +7292,134 @@ unittest
+    static if (T.sizeof > 4)
+        x |= x >> 32;
+
+    if (val < 0)
```
x++, Ditto, plus copysign from val to xx for floatingpoints. Excuse me for partial reviews.

JackStouffer replied:
>plus copysign from val to x for floating points

`copysign` is already being used in the floating point version, and you're the one who wrote it ;)

JackStouffer replied:
Fixed the `x++` thing

9il replied:
static if for signed?

9il replied:
x++;
static if()
  if
      x = -x;
return x;

JackStouffer replied:
Ah, sorry I misunderstood you.

JackStouffer commented:
@9il Fixed.

9il commented:
LGTM, ping @CyberShadow

9il commented:
> What about bsr?

From `std.uni`
```D
size_t floorPowerOf2(size_t arg) @safe pure nothrow @nogc
{
    import core.bitop : bsr;
    assert(arg > 1); // else bsr is undefined
    return 1<<bsr(arg-1);
}

size_t ceilPowerOf2(size_t arg) @safe pure nothrow @nogc
{
    import core.bitop : bsr;
    assert(arg > 1); // else bsr is undefined
    return 1<<bsr(arg-1)+1;
}
```

Hmm, it looks faster...

CyberShadow commented:
> LGTM, ping @CyberShadow

What is needed of me?

CyberShadow commented:
I proposed trying `bsr` way up in [comment #5](https://github.com/D-Programming-Language/phobos/pull/3755#issuecomment-149597521).

But please benchmark it, some of these old magic "one-opcode" instructions are no longer fast on modern CPUs, something about since they were not being used much in practice any more they are executed by some roundabout way of slow emulation.

9il commented:
> What is needed of me?

Push the "Merge" button. However looks like there is better `bsr`based  implementation from `std.uni`. I have found code 3 minutes ago.



> But please benchmark it, some of these old magic "one-opcode" instructions are no longer fast on modern CPUs, something about since they were not being used much in practice any more they are executed by some roundabout way of slow emulation.

Looks reasonable.

JackStouffer commented:
This code is twice as fast as the original integral version

```
T nextPow2(T)(const T val) if (isIntegral!T)
{
    import core.bitop : bsr;

    Unqual!T x = val;

    static if (isSigned!T) x = abs(x);

    x = 1 << bsr(x) + 1;

    static if (isSigned!T)
    if (val < 0)
        return -x;

    return x;
}
```

Working now to get it to pass all tests.

JackStouffer commented:
The above code changes overflow/underflow behavior. What is the preferred result?

CyberShadow commented:
> The above code changes overflow/underflow behavior. What is the preferred result?

Define overflow behavior to be what the simplest implementation returns.

Speaking of which, is `bsr` guaranteed to have the same behavior on all platforms?

JackStouffer commented:
Added bsr implementation

> Speaking of which, is bsr guaranteed to have the same behavior on all platforms?

Depends on druntime, so I'd assume so, but I don't know.

9il commented:
After all looks like that we have implementation for `floorPow2` both for integer and fp version. So I suggest remove `...+1` and rename.

9il commented:
bsr is not defined for x=0

JackStouffer commented:
> So I suggest remove ...+1 and rename.

I don't understand what you're trying to say here

JackStouffer commented:
> bsr is not defined for x=0

Added `if` statement for that case.

9il commented:
>> So I suggest remove ...+1 and rename.

> I don't understand what you're trying to say here

Integer implementation `bsr(x) + 1` and floating `exp + 1` can be reduced to `bsr(x)` and `exp`. So we can name functionality `floorPow2`.

9il commented on std/math.d:
```diff
@@ -7291,3 +7292,141 @@ unittest
+        x = abs(x);
+
+    if (x == 0)
+        return 0;
```
Tiny optimization:
```D
   if (x < 0)
       x = abs(x);

   if (x == 0)
       return x;
```

9il commented:
> Integer implementation bsr(x) + 1 and floating exp + 1 can be reduced to bsr(x) and exp. So we can name functionality floorPow2.

`truncPow2` is more precise definition.

JackStouffer commented:
> truncPow2 is more precise definition.

`trunc` as in truncation? Technically you are cutting off a zero from the start of the binary, not to be pedantic but when something is truncated it's shorter than what it was originally while here the resulting number is still the same number of bits. `nextPow2` is more descriptive IMO.

9il commented:
After removing `+1`s it would be shorter.

JackStouffer commented:
OH, I see what you're saying. Shouldn't `truncPow2` be a seperate function though? Why not have both `truncPow2` and `nextPow2` so you get both `floor` and `ceil` equivalents?

9il commented:
> OH, I see what you're saying. Shouldn't truncPow2 be a seperate function though? Why not have both truncPow2 and nextPow2 so you get both floor and ceil equivalents?

Looks good to have both.

9il commented:
For 32 bit platforms `bsr` should be called simulated for ulong. `bsr64` already available in `std.numeric`. Please move `bsr64 `to  `std.math` and make it private.

9il commented:
> Please move bsr64to std.math and make it private.

EDIT: `package`

9il commented:
Oh, looks like `bsr_ulong` already implemented in std.math. So we can remove `bsr64` from `std.numeric`.

JackStouffer commented:
Seems like `bsr_ulong` should be public and documented, no?

9il commented:
> Seems like bsr_ulong should be public and documented, no?

Only `package`. `bsr` is just assembler stuff.

JackStouffer commented:
* added truncPow2
* made bsr_ulong `package`

JackStouffer commented:
ping

9il commented on std/math.d:
```diff
@@ -7291,3 +7292,248 @@ unittest
+    static if (type == PowType.ceil)
+        x = Unqual!T(1) << bsr_ulong(x) + 1;
+    else
+        x = Unqual!T(1) << bsr_ulong(x);
```
bsr_ulong should be used only for ulong. For other types please use common bsr

JackStouffer commented:
@9il fixed

9il commented:
final LGTM. Please merge this PR.

JackStouffer commented:
ping @CyberShadow @DmitryOlshansky

CyberShadow commented:
Strange, I don't understand where some of the doc autotester changes are coming from.

9il commented:
@JackStouffer BTW, does it work with `ubyte`, `ushort`, `byte`, `short`?

JackStouffer commented:
@9il no, they don't seem to, currently investigating

JackStouffer commented:
Fixed ubyte, ushort, byte, short issues and fixed merge conflicts. I also changed this PR to refelect the fact that there now exists a ulong version of bsr.

JackStouffer commented:
@9il ping

JackStouffer commented:
ping @9il @DmitryOlshansky

DmitryOlshansky commented:
LGTM
@9il ?

9il commented:
Auto-merge toggled on

JackStouffer commented:
Thanks!

rainers commented on std/math.d:
```diff
@@ -7318,3 +7319,286 @@ unittest
+    assert(nextPow2(-2) == -4);
+    assert(nextPow2(-10) == -16);
+
+    assert(nextPow2(uint.max) == 1);
```
@JackStouffer Are these results of nextPow2 for uint/int.min/max actually desired behaviour? It isn't mentioned in the documentation and seems like an implementation artefact, i.e. it relies on `(1 << 33) == 1` which is undefined behaviour.
I stumbled across this while trying to get the tests pass with LDC, but these fail with optimizations (it eliminates the shift).

JackStouffer replied:
These are not really desired behavior, it's supposed to be documenting what happens with overflow and underflow.

When I asked about the strange behavior, @CyberShadow said

> Define overflow behavior to be what the simplest implementation returns.

So that's what I did.

rainers replied:
Thanks for the explanation. Unfortunately the current implementation depends on processor behaviour.
I think it would be more consistent and reliable to return 0 for the "out of bounds" cases, similar to what happens for floating point values. This should just add one more comparison.

JackStouffer replied:
https://issues.dlang.org/show_bug.cgi?id=15973

JackStouffer replied:
@rainers https://github.com/dlang/phobos/pull/4268
